romaji fails to build on beta,

[romaji](https://crater-reports.s3.amazonaws.com/beta-1.31-1/beta-2018-10-30/reg/romaji-0.1.1/log.txt), cc @uzimith

With a few minutes of time invested I wasn't able to narrow this down to a smaller example that still succeeded on stable. 

```
! 
Nov 02 07:32:11.418 INFO kablam! error[E0282]: type annotations needed
Nov 02 07:32:11.418 INFO kablam!   --> src/romaji.rs:80:24
Nov 02 07:32:11.418 INFO kablam!    |
Nov 02 07:32:11.418 INFO kablam! 80 |                 x if x.is_youon() && buffer.is_katakana() => {
Nov 02 07:32:11.419 INFO kablam!    |                        ^^^^^^^^ cannot infer type for `T`
Nov 02 07:32:11.419 INFO kablam!    |
Nov 02 07:32:11.419 INFO kablam!    = note: type must be known at this point
```

cc @rust-lang/libs

--- a/src/librustc_typeck/check/_match.rs
+++ b/src/librustc_typeck/check/_match.rs
@@ -626,9 +626,9 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
         let discrim_diverges = self.diverges.get();
         self.diverges.set(Diverges::Maybe);
 
-        // Typecheck the patterns first, so that we get types for all the
-        // bindings.
-        let all_arm_pats_diverge = arms.iter().map(|arm| {
+        // rust-lang/rust#55810: Typecheck patterns first (via eager
+        // collection into `Vec`), so we get types for all bindings.
+        let all_arm_pats_diverge: Vec<_> = arms.iter().map(|arm| {
             let mut all_pats_diverge = Diverges::WarnedAlways;
             for p in &arm.pats {
                 self.diverges.set(Diverges::Maybe);
@@ -644,7 +644,7 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
                 Diverges::Maybe => Diverges::Maybe,
                 Diverges::Always | Diverges::WarnedAlways => Diverges::WarnedAlways,
             }
-        });
+        }).collect();
 
         // Now typecheck the blocks.
         //

--- a/src/test/ui/typeck/issue-55810-must-typeck-match-pats-before-guards.rs
+++ b/src/test/ui/typeck/issue-55810-must-typeck-match-pats-before-guards.rs
@@ -1,0 +1,23 @@
+// compile-pass
+
+// rust-lang/rust#55810: types for a binding in a match arm can be
+// inferred from arms that come later in the match.
+
+struct S;
+
+impl S {
+    fn method(&self) -> bool {
+        unimplemented!()
+    }
+}
+
+fn get<T>() -> T {
+    unimplemented!()
+}
+
+fn main() {
+    match get() {
+        x if x.method() => {}
+        &S => {}
+    }
+}

